### ISSUE TITLE
wasm32 support

### ISSUE THREAD

**Issue by BartAdv** [slide-rs/specs#341]
Hello,

Tried to build it with `--target wasm32-unknown-unknown`, haven't succeeeded. At first, I found out the dispatcher might be problematic (it's supporting parallel execution), so I went and annotated `shred` with conditional compilation directives (it already had it for `emscripted`, so I just duplicated it with analogue for `target_arch = "wasm32"` [here](https://github.com/bartadv/shred/tree/wasm32). It builds fine then.

Going back to `specs`, I've overriden the `shred` to use my branch:

```
[dependencies.shred]
git = "https://github.com/BartAdv/shred"
rev = "4d6b1a9c4d0bc1cd8d7bba275e7fbb4d3be69f1b"
```

only to see

```
error[E0433]: failed to resolve. Use of undeclared type or module `imp`
  --> /home/bart/.cargo/registry/src/github.com-1ecc6299db9ec823/rand-0.3.20/src/os.rs:35:18
   |
35 | pub struct OsRng(imp::OsRng);
   |                  ^^^ Use of undeclared type or module `imp`

error[E0433]: failed to resolve. Use of undeclared type or module `imp`
  --> /home/bart/.cargo/registry/src/github.com-1ecc6299db9ec823/rand-0.3.20/src/os.rs:40:9
   |
40 |         imp::OsRng::new().map(OsRng)
   |         ^^^ Use of undeclared type or module `imp`

```

OK, I get that - rand crate relies on some sys stuff. So I wanted to fix whatever dep is wanting `rand`. Found out, the `rayon-core` uses it. As it turns out, `rayon` is used by `hibitset` (gated by feature flag `parallel`), so I wanted to check if I could gate the specs rayon dependncy similarily...:

```
diff --git a/Cargo.toml b/Cargo.toml
index ef2a182..a979465 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -23,10 +23,9 @@ derivative = "1"
 fnv = "1.0"
 hibitset = { version = "0.4.1", features = ["parallel"] }
 mopa = "0.2"
-shred = "0.5.0"
 shred-derive = "0.3"
 tuple_utils = "0.2"
-rayon = "0.8.2"
+rayon = { version = "0.8.2", optional = true}
 
 futures = { version = "0.1", optional = true }
 serde = { version = "1.0", optional = true, features = ["serde_derive"] }
@@ -34,9 +33,14 @@ serde = { version = "1.0", optional = true, features = ["serde_derive"] }
 # enable rudy via --features rudy
 rudy = { version = "0.1", optional = true }
 
+[dependencies.shred]
+git = "https://github.com/BartAdv/shred"
+rev = "4d6b1a9c4d0bc1cd8d7bba275e7fbb4d3be69f1b"
+
 [features]
 common = ["futures"]
 nightly = []
+parallel = ["hibitset/parallel", "rayon"]
 
 [package.metadata.docs.rs]
 features = ["common", "serde"]
```

...to no avail, the build is still giving me

```
error[E0433]: failed to resolve. Use of undeclared type or module `imp`
  --> /home/bart/.cargo/registry/src/github.com-1ecc6299db9ec823/rand-0.3.20/src/os.rs:35:18
   |
35 | pub struct OsRng(imp::OsRng);
   |                  ^^^ Use of undeclared type or module `imp`
```

Now - maybe someone more compoenent than me could try and check it? Is it planned to support `wasm32-unknown-unknown` target?

**Comment by BartAdv** [comment]
OK, I've learnt bit more about those things, so I am able to proceed with that - I'm wondering if I could clean up a bit how this multithreading support is conditioned. As I currently see, the `shred` detects whether we build for wasm32, and disables with that, whereas `hibitset` uses `parallel` feature flag. `specs` doesn't use it, so it could be added, just wondering what's better approach...

**Comment by BartAdv** [comment]
Got the WIP here: https://github.com/BartAdv/specs/commit/dd6052b165c7420b2401e408c96218c658bf7a0e

**Comment by BartAdv** [comment]
Since the discussion got a bit spread amongst repos/PRs, I want to step back a bit, and decide what would be best approach:

1) extract parallel features from specs into specs-parallel?
2) conditionally build with parallel support with `parallel` feature

On top of that: is it really only about parallel? `shred` depends on `pulse`, which is for 'async signals'. Can it be put under the same umbrella as parallel? When it comes to wasm32, you can't have thread, but I'm not sure about some async stuff...

**Comment by torkleyy** [comment]
Yeah, let's just put it all under a `parallel` flag. I'm wondering, how long will it take for wasm to support multi-threading?

**Comment by torkleyy** [comment]
@BartAdv Are you still interested in this?

**Comment by BartAdv** [comment]
@torkleyy yes! It slipped off my radar while waiting for shred changes to be merged. I think I got all the bits ready somewhere, because I'm already building a project using specs with wasm, so I'll just check what's needed to finalize it, most likely later this day.

**Comment by torkleyy** [comment]
Great, thank you!

**Comment by BartAdv** [comment]
@torkleyy OK, so it seems to be all ready in [this](https://github.com/bartadv/specs/tree/wasm32) branch, but it references shred via git hash, so I guess first we need shred release?

**Comment by torkleyy** [comment]
shred had a release lately, there's an open PR that bumps the shred version.

**Comment by BartAdv** [comment]
OK, ping me when it's done, I'll rebase on top of that then.

**Comment by torkleyy** [comment]
@BartAdv It's merged ;)

**Comment by BartAdv** [comment]
Yep, sorry for delay, I've opened the PR #353 for that

**Comment by richardanaya** [comment]
Any updates on this? :) I keep poking my head into this issue. It sounds like shred is wasm compliant now? What major pieces remain.

**Comment by willglynn** [comment]
Status links:

* Master tracking issue: https://github.com/WebAssembly/design/issues/1073
* Firefox: [#1389458](https://bugzilla.mozilla.org/show_bug.cgi?id=1389458)
* Edge: https://github.com/Microsoft/ChakraCore/issues/3477
* Chrome: [feature dashboard](https://www.chromestatus.com/features/5724132452859904)

A related issue is [`SharedArrayBuffer` browser support](https://caniuse.com/#feat=sharedarraybuffer), which was yanked by everybody pending Spectre/Meltdown mitigations, and was just recently re-enabled in Chrome 68 due to some [significant browser memory model changes](https://security.googleblog.com/2018/07/mitigating-spectre-with-site-isolation.html). This work suggests a shorter path for getting `wasm32` threads in Chrome, and a longer path for everyone else.

**Comment by bgourlie** [comment]
For what it's worth, I was able to get get the clusterbomb example compiled and working in the browser. At a high level I had to:

- Pull in specs as a git dependency since [this commit](https://github.com/slide-rs/specs/commit/97a74c89a33aafca4a1d602292ce6bfa3239c4d2) is not yet released
- set `default-features = false` for specs in my `Cargo.toml`
- Depend on `rand` with feature `wasm-bindgen`
- Change `durability_range.sample(..)` to `rng.gen_range(..)` and get rid of any references to `Range`
- Change all `par_join` to `join`
- Create `console_log!` macro and use that instead of `println!`

**Comment by bgourlie** [comment]
Wanted to give a quick update on this. There are still challenges unrelated to getting specs to compile to wasm that stand in the way of using specs for anything non-trivial in the browser. At a high-level, constraints enforced by wasm-bindgen and the nature of game loops in javascript are proving difficult to work around. I've outlined the challenges in this [stackoverflow post](https://stackoverflow.com/questions/53000413/how-can-i-work-around-not-being-able-to-export-functions-with-lifetimes-when-usi).

**Comment by liamcurry** [comment]
Any progress/plans on this?

**Comment by torkleyy** [comment]
@liamcurry This is mostly blocked on `rayon`; there are two options. We could either implement wasm support for rayon, or implement a fallback for wasm.

Specs should be usable with wasm (with some challanges noted by @bgourlie) and missing support for `par_join` (blocked on rayon).

This is not a must-have for 1.0.

**Comment by torkleyy** [comment]
This is still relevant.

**Comment by erlend-sh** [comment]
Relevant `rayon` issue: https://github.com/rayon-rs/rayon/pull/636

**Comment by torkleyy** [comment]
Initial support has landed (non-parallel), #580 tracks further progress.
